feat: enable global beta features through DB

LinHuiqing opened the pull request:
## Problem
<!-- What problem are you trying to solve? What issue does this close? -->

Manually adding beta flags for individual users is unnecessarily tedious in non-production environments.

## Solution
<!-- How did you solve the problem? -->

Create a new DB collection to store global beta flag statuses.

**Breaking Changes** 
<!-- Does this PR contain any backward incompatible changes? If so, what are they and should there be special considerations for release? 
- [ ] Yes - this PR contains breaking changes
    - Details ... -->
- [x] No - this PR is backwards compatible  

**Features**:

- Create new `featureflags` DB collection
- Show beta features if either user or global beta flag is set to enabled; do not show beta features if both flags are not enabled

**Bug Fixes**:

- Rename `add-payment-flag.js` and `multi-language-stats` scripts' directory 

## Tests
<!-- What tests should be run to confirm functionality? -->

Currently, the only beta feature is payments, so we will be basing the tests on the payment feature.

1. Ensure that users with no beta flag + no global beta flag enabled will not see beta features:
- [ ] On the DB, check that:
   - [ ] In the `user` collection, your user document, `betaFlags: { payment: false }`.
   - [ ] In the `globalBeta` collection, a document has the following values `{ name: payment, enabled: false }`.
- [ ] Create a new storage mode form. Go to the settings tab of a storage form, there should not be a "Payments" drawer.
- [ ] Go to the create tab of a storage form, there should not be a "Payments" tab in the builder drawer.
- [ ] Payment preview should not be shown in:
   - [ ] Builder view
   - [ ] Admin preview
   - [ ] Public form

2. Ensure that all users will see beta feature if global beta flag is enabled:
- [ ] On the DB, create the `globalBeta` collection if it doesn't exist. Add a document `{ name: payment, enabled: true }`.
- [ ] On the DB, go to the `user` collection, find your user document, and set `payment` beta flag to `false`.
- [ ] Go to the settings tab of a storage form, there should be a "Payments" drawer.
   - [ ] Connect to a Stripe account.
- [ ] Go to the create tab of a storage form, there should be a "Payments" tab in the builder drawer.
   - [ ] Enable payments, edit name and payment amount. Save.
   - [ ] Payment preview should reflect saved changes accordingly.
- [ ] Form preview should show payment preview accordingly.
- [ ] Public form should:
   - [ ] Show payment preview accordingly.
   - [ ] Allow payment successfully.

3. Ensure that users with user beta flag will still see beta feature if global beta flag is enabled:
- [ ] On the DB, go to the `globalBeta` collection, find the document with name `payment` and make sure the `enabled` value is `true`.
- [ ] On the DB, go to the `user` collection, find your user document, and set `payment` beta flag to `true`.
- [ ] Go to the settings tab of a storage form, there should be a "Payments" drawer.
   - [ ] Connect to a Stripe account.
- [ ] Go to the create tab of a storage form, there should be a "Payments" tab in the builder drawer.
   - [ ] Enable payments, edit name and payment amount. Save.
   - [ ] Payment preview should reflect saved changes accordingly.
- [ ] Form preview should show payment preview accordingly.
- [ ] Public form should:
   - [ ] Show payment preview accordingly.
   - [ ] Allow payment successfully.

4. Ensure that if global beta flag is not enabled, users with user beta flag will still see beta feature:
- [ ] On the DB, go to the `globalBeta` collection, find the document with name `payment` and set the `enabled` value to `false`.
- [ ] On the DB, go to the `user` collection, find your user document, and set `payment` beta flag to `true`.
- [ ] Go to the settings tab of a storage form, there should be a "Payments" drawer.
   - [ ] Connect to a Stripe account.
- [ ] Go to the create tab of a storage form, there should be a "Payments" tab in the builder drawer.
   - [ ] Enable payments, edit name and payment amount. Save.
   - [ ] Payment preview should reflect saved changes accordingly.
- [ ] Form preview should show payment preview accordingly.
- [ ] Public form should:
   - [ ] Show payment preview accordingly.
   - [ ] Allow payment successfully.

## Deploy Notes
<!-- Notes regarding deployment of the contained body of work.  -->
<!-- These should note any new dependencies, new scripts, etc. -->

**New scripts**:

- `scripts/20230508_add-feature-flag/add-feature-flag-collection.js`: run in Studio3T for new `featureflags` collection to be added

LinHuiqing committed:
chore: correct date in add-payment-flag filename

LinHuiqing committed:
feat: get global beta flags from db

LinHuiqing committed:
chore: correct date in multi-language-stats filename

LinHuiqing committed:
fix: allow payment requests if global beta enabled

LinHuiqing committed:
refactor: move global beta flags to admin form

LinHuiqing committed:
refactor: set up beta flag shared constants

LinHuiqing committed:
chore: update comments for global beta flag code

timotheeg commented on scripts/20230508_add-global-beta/add-global-beta-collection.js:
```diff
@@ -0,0 +1,20 @@
+  { name: 'payment' },
+  {
+    $setOnInsert: {
+      enabled: true,
```
Generally creation values for beta flags should be `false`, and turning something on should be an explicit separate action.

Like, we need to make sure we do not run this script in production right now 😰

LinHuiqing replied:
Oh yes, that's true! Sorry about this :')

LinHuiqing replied:
Done in [850b3e5](https://github.com/opengovsg/FormSG/pull/6286/commits/850b3e52f18c4c57fa5b3f67e253d263faa68e16)

timotheeg commented on frontend/src/features/env/queries.ts:
```diff
@@ -43,3 +46,15 @@ export const usePublicFeedbackFormView = (
 }
+
+type UseGlobalBetaReturn = {
+  flagEnabled?: boolean
```
I think we should drop the prefix `flag`, it's clear from the function name and context that we're getting a flag by name, so just `enabled` should do.

LinHuiqing replied:
Done in [d8c8e1b](https://github.com/opengovsg/FormSG/pull/6286/commits/d8c8e1b640c2e0be856bd833dfc48d9a43ad09f8)

timotheeg commented on frontend/src/features/admin-form/settings/SettingsPage.tsx:
```diff
@@ -106,7 +110,7 @@ export const SettingsPage = (): JSX.Element => {
             <SettingsWebhooksPage />
           </TabPanel>
-          {user?.betaFlags?.payment && (
+          {(user?.betaFlags?.payment || paymentGlobalBeta) && (
```
Since this is duplicated, let's introduce a reusable constant `displayPayments` like it's done in [`FieldListDrawer.tsx`](https://github.com/opengovsg/FormSG/pull/6286/files#diff-2f4c1d0c5a445b55c5fdf1d5d4dc5bde69c281a75e4328f98478560ae7508f6c)

LinHuiqing replied:
Done in [c9be145](https://github.com/opengovsg/FormSG/pull/6286/commits/c9be145a81e1b1ef683e137debc7bdea73c66a5b)

timotheeg commented on src/app/modules/form/admin-form/admin-form.service.ts:
```diff
@@ -1610,3 +1613,24 @@ export const updatePayments = (
+        error,
+      })
+
+      return new DatabaseError(
+        `Unable to get global beta flag status. If this issue persists, please submit a Support Form at (${SUPPORT_FORM_LINK})`,
+      )
```
Hmm, throwing a runtime error is a bit extreme. I'm thinking if we should return `false`, and make sure we have an alert in place when we detect we cannot read the flags.

LinHuiqing replied:
Decided to keep this DatabaseError and leave the decision of what to return to controllers. Controllers now treat DatabaseErrors the same as returning false, done in [f537622](https://github.com/opengovsg/FormSG/pull/6286/commits/f537622f7ed93d0f01cbc7a4bb0139f0980c45a6).

timotheeg commented on src/app/modules/form/admin-form/admin-form.utils.ts:
```diff
@@ -501,9 +501,10 @@ export const verifyValidUnicodeString = (value: any, helpers: any) => {
   betaFlag: keyof Exclude<IUserSchema['betaFlags'], undefined>,
 ) => {
-  return user?.betaFlags?.[betaFlag]
+  return user?.betaFlags?.[betaFlag] || globalUserBeta
```
Thinking about this again, I think the naming is wrong. TYhe global beta flag actually activates the feature for the whole platform, to it's really a feature flag, not a beta flag. 

Beta does work for a group of users or agencies, so having beta flags in user objects, or in agency objects works, but global beta sounds weird 🤔 

Just for terminology completeness, the way the global flags are implemented now is indeed not a kill switch, because turning them off still lets the beta user use it.

(Not Related to PR: I'm thinking we should actually discuss kill switches for FormSG at some point. With feature like payment, it makes sense to me to be able to turn all payment off if something catastrophic happens 🤔 )

LinHuiqing replied:
@timotheeg so for this, should all the related names be "global feature flag" instead of "global beta flag"?

timotheeg replied:
Yes Mam'! 😃

timotheeg replied:
In fact, "feature flag" is probably enough

LinHuiqing replied:
Done in [b2dd2f3](https://github.com/opengovsg/FormSG/pull/6286/commits/b2dd2f3b5175d5c82b91039d2fbbf0928eb7840d), [90362ee](https://github.com/opengovsg/FormSG/pull/6286/commits/90362eee30347c1582b1bdb49bb73e886a2a4a91) and [78f8b6b](https://github.com/opengovsg/FormSG/pull/6286/commits/78f8b6bfe2f5f7710152aa3014fc1a619ae9fa5e).

timotheeg reviewed (commented):
Thanks HuiQing. I have some comments. Feels a bit like I'm bike-shedding, but I think we need to be a bit stricter with terminology.

KenLSM commented on src/app/routes/api/v3/admin/forms/admin-forms.routes.ts:
```diff
@@ -32,3 +33,4 @@ AdminFormsRouter.use(AdminFormsPresignRouter)
 AdminFormsRouter.use(AdminFormsLogicRouter)
 AdminFormsRouter.use(AdminFormsTwilioRouter)
 AdminFormsRouter.use(AdminFormsPaymentsRouter)
+AdminFormsRouter.use(AdminFormsBetaRouter)
```
Should global feature flags be tied to `admins`? Would it make sense to be a service by itself? This would allow for optimization relating to caching feature flag routes, and/ or filter errors by routes.

timotheeg replied:
I had a similar thought.

Feature flags for respondents could even be served alongside the environment dataset the app gets on bootstrap 🤔 

It probably still makes sense to still have some beta flag endpoint for admins, and the logic can be server side to take everything into account: personal+global flags, or whatever other criteria we chose (e.g. membership to agency?)

That's beyond scope for this PR though, but I do think if the PR introduces global flags, we can serve them from an independent endpoint (not scoped to admins)

LinHuiqing replied:
okays, i'll serve it from an independent endpoint, should be relatively quick work 👍

LinHuiqing replied:
Done here: [2d28cc7](https://github.com/opengovsg/FormSG/pull/6286/commits/2d28cc70c26eaab7b943a14d373bb15696a583f9)

LinHuiqing committed:
fix: set global payment beta flag to false

LinHuiqing committed:
refactor: global beta schema definition into compile

LinHuiqing committed:
refactor: rename flagEnabled to enabled

LinHuiqing committed:
refactor: reusable displayPayments constant

LinHuiqing committed:
fix: return false instead of throw err for getGlobalBeta

- can decide to move returning false to service level in the future if we decide otherwise

LinHuiqing committed:
refactor: rename global beta to feature flags

LinHuiqing committed:
refactor: rename global beta to feature flag for script

LinHuiqing committed:
refactor: rename beta-flags as feature-flags in shared constants

timotheeg commented on scripts/20230508_add-feature-flag/add-feature-flag-collection.js:
```diff
@@ -0,0 +1,20 @@
+      enabled: false,
+    },
+  },
+  { upsert: true },
```
Is there an option for "on collision do nothing"?

We don't have to upsert any value and risk overwriting the environment 🤔

LinHuiqing replied:
This already does it!

timotheeg commented on src/app/modules/form/admin-form/admin-form.payments.controller.ts:
```diff
@@ -50,11 +51,36 @@ export const handleConnectAccount: ControllerHandler<{
+    logger.error({
+      message: 'Error occurred whilst retrieving feature flag status',
+      meta: logMeta,
+      error: featureFlagEnabledResult.error,
```
For clarity, add the name of the flag here in the log.

LinHuiqing replied:
Done in [f879381](https://github.com/opengovsg/FormSG/pull/6286/commits/f87938197b3492a6014faab219137370eb202cd9)

timotheeg commented on src/app/modules/form/admin-form/admin-form.utils.ts:
```diff
@@ -501,9 +501,10 @@ export const verifyValidUnicodeString = (value: any, helpers: any) => {
  */
 export const verifyUserBetaflag = (
   user: IUserSchema,
+  featureFlagEnabled: boolean,
```
This looks a bit funny. The function name indicate it cares for a user level beta flag. If the caller wants to handle a feature flag, that should be its job 🤔 

I get you are trying to stay DRY because there are 2 calls, but I think it's better to keep the user beta check narrow and specific.

LinHuiqing replied:
done in [e8eb023](https://github.com/opengovsg/FormSG/pull/6286/commits/e8eb023d5a5db5f495414aeb6f0523e2db7d933a)

timotheeg commented on frontend/src/features/admin-form/common/queries.ts:
```diff
@@ -154,3 +155,9 @@ export const useFormTemplate = (
+
+export const getFeatureFlag = async (flag: string): Promise<boolean> => {
+  return ApiService.get<boolean>('/admin/forms/feature-flag', {
+    params: { flag },
```
I think it's a bit funny that the FE makes requests for one feature flag at a time. I reckon that typically, server would supply all the active feature flags, and the FE will default everything not supplied to `false`.

LinHuiqing replied:
hmm okay makes sense, so we return a list of all active feature flags (and keep this list small)

LinHuiqing replied:
done in [74fef15](https://github.com/opengovsg/FormSG/pull/6286/commits/74fef152031d00e94004e4d539622235363b1206) and [f879381](https://github.com/opengovsg/FormSG/pull/6286/commits/f87938197b3492a6014faab219137370eb202cd9)

timotheeg commented on src/app/modules/form/admin-form/admin-form.payments.controller.ts:
```diff
@@ -212,15 +235,43 @@ export const _handleUpdatePayments: ControllerHandler<
       // Step 2: Check if user has 'payment' betaflag
-      .andThen((user) => verifyUserBetaflag(user, 'payment'))
+      .andThen((user) =>
+        verifyUserBetaflag(user, featureFlagEnabled, featureFlags.payment),
```
We can keep the logic here:
```typescript
return featureFlagEnabled ? ok(user) : verifyUserBetaflag(user, 'payment')
```

And on the other instance line 82 as well.

LinHuiqing replied:
done in [e8eb023](https://github.com/opengovsg/FormSG/pull/6286/commits/e8eb023d5a5db5f495414aeb6f0523e2db7d933a)

LinHuiqing committed:
refactor: separate endpoint for feature flags

LinHuiqing committed:
feat: get all enabled feature flags instead

LinHuiqing committed:
refactor: rm feature flag check in verifyUserBetaFlag

LinHuiqing committed:
fix: remove unnecessary params in handleGetEnabledFlags

- update error messages for controllers using getEnabledFeatureFlags too

LinHuiqing commented on src/app/modules/feature-flags/feature-flags.controller.ts:
```diff
@@ -0,0 +1,38 @@
+ */
+export const handleGetEnabledFlags: ControllerHandler<
+  never,
+  // TODO: stricter typing to restrict typing to flag values in shared/constants
+  string[] | ErrorDto
```
Should this typing be an important change before we deploy it? 🤔 or any tips? I was thinking we could type it as `"payment" | "flag1" | "flag2"[]` based on what is set in shared/constants

timotheeg replied:
No, I think we can leave the TODO in for now.

timotheeg commented on src/app/modules/feature-flags/feature-flags.service.ts:
```diff
@@ -0,0 +1,25 @@
+      error,
+    })
+
+    return new DatabaseError(`Unable to get feature flag status.`)
```
```suggestion
    return new DatabaseError(`Unable to fetch enabled feature flags.`)
```

LinHuiqing replied:
done in [8eac35d](https://github.com/opengovsg/FormSG/pull/6286/commits/8eac35d990fbccec26471879e85e014e5754a405)

timotheeg commented on frontend/src/features/feature-flags/queries.ts:
```diff
@@ -0,0 +1,11 @@
+  base: ['feature-flags'] as const,
+}
+
+export const useFeatureFlagsEnabled = (): UseQueryResult<Set<string>> => {
```
Why did you have `enabled` as a suffix rather than a prefix like on the server? It reads a little funny like this 🤔 😅 

I'm thinking that in the API and in frontend code, `enabled` is not needed. The code is self explanatory like this:

```typescript
const flags = useFeatureFlags()

if (flags.has('payments')) {
    // do stuff
}
```

LinHuiqing replied:
Done in [bba411c](https://github.com/opengovsg/FormSG/pull/6286/commits/bba411c888f1dff11054ea64a008ba9e3e7f40a1)

timotheeg commented on src/app/modules/feature-flags/feature-flags.service.ts:
```diff
@@ -0,0 +1,25 @@
+const logger = createLoggerWithLabel(module)
+const FeatureFlagModel = getFeatureFlagModel(mongoose)
+
+export const getEnabledFlags = (): ResultAsync<string[], DatabaseError> => {
```
Please add a `TODO` that eventually we likely need a local caching system for feature flags, with some interval based refresh. That's because querying DB every time should be unnecessary for infrequently changing resources like feature flags.

You could check how we did it for OIDC remote NDI conf (cc @tshuli)
1) load on bootstrap (and fail to start if unable to fetch)
2) use local copy for all subsequent uses
3) set up a interval-based refresh, if failure to refesh: log but keep using stale data (no runtime or user visible errors)

I'm actually wondering if we need to have that caching system from day 1 for feature flags because I have a little worry that we are possibly the number of requests to DB quite a bit 🤔 

In my initial thought for a feature flag system, we'd just use env vars, meaning the feature flag value would be static for the entire lifetime of the process. Even if the feature flags come from DB now, doing a singular boostrap-time load is probably enough for now, and we can always do an env restart to force a refetch if needed.

LinHuiqing replied:
TODO added in [06c7f79](https://github.com/opengovsg/FormSG/pull/6286/commits/06c7f792af2967f467d27eecdfd0abd66677da14) and [e64cba8](https://github.com/opengovsg/FormSG/pull/6286/commits/e64cba8ebf494c3c07526db8021284df52c459e8)

LinHuiqing committed:
chore: update db err message for getEnabledFlags

LinHuiqing committed:
chore: feature flags nomenclature

LinHuiqing committed:
chore: add TODO for local caching system

LinHuiqing committed:
chore: link to comment instead of PR

timotheeg reviewed (approved):
LGTM

LinHuiqing committed:
chore: update new collection name
